lib/testutils/resolver.go

```diff
@@ -0,0 +1,228 @@
+/*
+ *
+ * k6 - a next-generation load testing tool
+ * Copyright (C) 2020 Load Impact
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package testutils
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/davecgh/go-spew/spew"
+	"golang.org/x/net/dns/dnsmessage"
+)
+
+// MockResolver implements netext.DNSResolver, and allows changing the host
+// mapping at runtime.
+type MockResolver struct {
+	m     sync.RWMutex
+	hosts map[string]net.IP
+}
+
+func NewMockResolver(hosts map[string]net.IP) *MockResolver {
+	if hosts == nil {
+		hosts = make(map[string]net.IP)
+	}
+	return &MockResolver{hosts: hosts}
+}
+
+func (r *MockResolver) Fetch(host string) (net.IP, error) {
+	r.m.RLock()
+	defer r.m.RUnlock()
+	if ip, ok := r.hosts[host]; ok {
+		return ip, nil
+	}
+	return nil, fmt.Errorf("lookup %s: no such host", host)
+}
+
+func (mr *MockResolver) handle(s net.Conn) {
+	for {
+		b := make([]byte, 512)
+		n, err := s.Read(b)
+		if err != nil {
+			return
+		}
+
+		var msg dnsmessage.Message
+		// FIXME: This returns the error `unpacking Question.Name: segment prefix is reserved`
+		// https://github.com/golang/net/blob/62affa334b73ec65ed44a326519ac12c421905e3/dns/dnsmessage/message.go#L1999
+		if err := msg.Unpack(b[:n]); err != nil {
+			spew.Dump(err)
```

Comment:
@na-- This is the test issue I'm having. I'm not sure why a wrong "segment prefix" is received here. I think it refers to the subnet prefix `/64` or `/128`, but a) the tests should be doing an IPv4 lookup not IPv6, and b) I don't see where this prefix is set in the request or what's actually wrong with the payload. :-/

Reply:
I'm missing big chunks of the background knowledge, but to be fair, I don't think this test is required for releasing the feature. I can be releases together with the rest of the DNS changes in v0.29.0

Reply:
Not an issue anymore since I abandoned this whole DNS mock server approach, as discussed.